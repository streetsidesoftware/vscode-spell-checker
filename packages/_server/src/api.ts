export * from './api/index.js';
