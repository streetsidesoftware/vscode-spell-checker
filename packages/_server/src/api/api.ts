import type {
    ApiPrefix,
    ApplyNotificationAPI,
    ApplyRequestAPI,
    ClientAPIDef,
    ClientSideMethods,
    Logger,
    MessageConnection,
    RpcAPI,
    ServerAPIDef,
    ServerSideMethods,
} from 'json-rpc-api';
import { createClientApi, createServerApi } from 'json-rpc-api';

import type {
    GetConfigurationForDocumentRequest,
    GetConfigurationForDocumentResult,
    IsSpellCheckEnabledResult,
    OnSpellCheckDocumentStep,
    PublishDiagnostics,
    SpellingSuggestionsResult,
    SplitTextIntoWordsResult,
    TextDocumentInfo,
    WorkspaceConfigForDocumentRequest,
    WorkspaceConfigForDocumentResponse,
} from './apiModels.js';

export type { Logger, MessageConnection } from 'json-rpc-api';

/** Requests that can be made to the server */
export interface ServerRequestsAPI {
    getConfigurationForDocument(req: GetConfigurationForDocumentRequest): GetConfigurationForDocumentResult;
    isSpellCheckEnabled(req: TextDocumentInfo): IsSpellCheckEnabledResult;
    splitTextIntoWords(req: string): SplitTextIntoWordsResult;
    spellingSuggestions(word: string, doc?: TextDocumentInfo): SpellingSuggestionsResult;
}

/** Notifications that can be sent to the server */
export interface ServerNotificationsAPI {
    notifyConfigChange: () => void;
    registerConfigurationFile: (path: string) => void;
}

/**
 * Requests that can be made from the server to the client(vscode extension)
 * Note: RPC requests to the client/extension is rare.
 */
export interface ClientRequestsAPI {
    onWorkspaceConfigForDocumentRequest: (req: WorkspaceConfigForDocumentRequest) => WorkspaceConfigForDocumentResponse;
}

/** Notifications from the server to the client(vscode extension) */
export interface ClientNotificationsAPI {
    onSpellCheckDocument(step: OnSpellCheckDocumentStep): void;
    onDiagnostics(pub: PublishDiagnostics): void;
}

export interface SpellCheckerServerAPI extends RpcAPI {
    serverRequests: ApplyRequestAPI<ServerRequestsAPI>;
    serverNotifications: ApplyNotificationAPI<ServerNotificationsAPI>;
    clientRequests: ApplyRequestAPI<ClientRequestsAPI>;
    clientNotifications: ApplyNotificationAPI<ClientNotificationsAPI>;
}

/**
 * Used on the server side  to communicate with the client(extension).
 */
export interface ServerSideApi extends ServerSideMethods<SpellCheckerServerAPI> {}
/**
 * Used in the client(extension) to communicate with the server.
 */
export interface ClientSideApi extends ClientSideMethods<SpellCheckerServerAPI> {}

export type ServerSideApiDef = ServerAPIDef<SpellCheckerServerAPI>;
export type ClientSideApiDef = ClientAPIDef<SpellCheckerServerAPI>;

export interface ServerSideHandlers {
    serverRequests: DefineHandlers<ServerSideApiDef['serverRequests']>;
    serverNotifications: DefineHandlers<ServerSideApiDef['serverNotifications']>;
}

// todo: make '' when all old apis are removed.
const pfx = '_';

const apiPrefix: ApiPrefix = {
    serverNotifications: pfx,
    serverRequests: pfx,
    clientNotifications: pfx,
    clientRequests: pfx,
};

export function createServerSideApi(
    connection: MessageConnection,
    api: ServerAPIDef<SpellCheckerServerAPI>,
    logger: Logger | undefined,
): ServerSideApi {
    return createServerApi(connection, api, logger, apiPrefix);
}

export function createClientSideApi(
    connection: MessageConnection,
    api: ClientAPIDef<SpellCheckerServerAPI>,
    logger: Logger | undefined,
): ClientSideApi {
    return createClientApi(connection, api, logger, apiPrefix);
}

type DefineHandlers<T> = {
    [P in keyof T]: Exclude<T[P], boolean>;
};
