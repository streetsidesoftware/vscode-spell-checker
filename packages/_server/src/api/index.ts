export * from './api.js';
// export * from './api.old.js';
export type * from './apiModels.js';
export * from './CommandsToClient.js';
export type * from './models/index.mjs';
export type { Suggestion } from './models/Suggestion.mjs';
