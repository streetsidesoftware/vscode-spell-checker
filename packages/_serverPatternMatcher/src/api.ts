import type { CSpellUserSettings } from '@cspell/cspell-types';
/**
 * Method signatures for requests to the Server.
 */
export type ServerRequestApi = {
    [key in keyof ServerMethods]: ApiReqResFn<ServerMethods[key]>;
};

/**
 * Internal Server Handler signatures to the Server API
 */
export type ServerRequestApiHandlers = ApiHandlers<ServerMethods>;

/**
 * Server RPC Request and Result types
 */
export interface ServerMethods {
    matchPatternsInDocument: ReqRes<MatchPatternsToDocumentRequest, MatchPatternsToDocumentResult>;
}

export type ServerRequestMethods = keyof ServerMethods;

export type ServerRequestMethodConstants = {
    [key in ServerRequestMethods]: key;
};

export interface TextDocumentRef {
    uri: UriString;
}

export interface NamedPattern {
    name: string;
    pattern: string | string[];
}

export interface PatternSettings {
    patterns: CSpellUserSettings['patterns'];
}

export interface MatchPatternsToDocumentRequest extends TextDocumentRef {
    patterns: (string | NamedPattern)[];
    settings: PatternSettings;
}

export type StartIndex = number;
export type EndIndex = number;

export type RangeTuple = [StartIndex, EndIndex];

export interface RegExpMatch {
    regexp: string;
    matches: RangeTuple[];
    elapsedTime: number;
    errorMessage?: string;
}

export type RegExpMatchResults = RegExpMatch;

export interface PatternMatch {
    name: string;
    defs: RegExpMatch[];
}

export interface MatchPatternsToDocumentResult {
    uri: UriString;
    version: number;
    patternMatches: PatternMatch[];
    message?: string;
}

export type UriString = string;
export type DocumentUri = UriString;

export type Req<T> = T extends { request: infer R } ? R : never;
export type Res<T> = T extends { response: infer R } ? R : never;
export type Fn<T> = T extends { fn: infer R } ? R : never;
export type OrPromise<T> = Promise<T> | T;

export interface ReqRes<Req, Res> {
    request: Req;
    response: Res;
}

/**
 * Utility type to combine the Request and Response to create the Handler function
 */
export interface RequestResponseFn<ReqRes> {
    request: Req<ReqRes>;
    response: Res<ReqRes>;
    fn: ApiReqHandler<ReqRes>;
}

export type ApiReqResFn<ReqRes> = ApiFn<Req<ReqRes>, Res<ReqRes>>;
export type ApiFn<Req, Res> = (req: Req) => Promise<Res>;

export type ApiReqHandler<ReqRes> = ApiHandler<Req<ReqRes>, Res<ReqRes>>;
export type ApiHandler<Req, Res> = (req: Req) => OrPromise<Res>;

export type ApiHandlers<ApiReqRes> = {
    [M in keyof ApiReqRes]: ApiReqHandler<ApiReqRes[M]>;
};

export type ApiReqResMethods<ApiReqRes> = {
    [M in keyof ApiReqRes]: ApiReqResFn<ApiReqRes[M]>;
};
