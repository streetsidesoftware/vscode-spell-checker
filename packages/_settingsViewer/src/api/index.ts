export * from './message';
export * from './MessageBus';
export * from './WebviewApi';
