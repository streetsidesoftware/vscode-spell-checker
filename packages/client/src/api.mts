import type * as API from 'code-spell-checker-server/api';

import * as di from './di.mjs';

export type DocumentInfo = API.TextDocumentInfo;
export type CheckDocumentOptions = API.CheckDocumentOptions;
export type CheckDocumentResponse = API.CheckDocumentResult;

export async function checkDocument(doc: DocumentInfo, options?: CheckDocumentOptions): Promise<CheckDocumentResponse> {
    const client = di.get('client');
    return client.serverApi.checkDocument(doc, options);
}
