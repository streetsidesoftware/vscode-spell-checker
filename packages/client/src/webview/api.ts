import { createDisposeMethodFromList, type DisposableLike, injectDisposable } from 'utils-disposables';
import { window } from 'vscode';
import { type MessageConnection } from 'vscode-jsonrpc/node';
import { setLogLevel } from 'vscode-webview-rpc/logger';
import { createServerSideHelloWorldApi } from 'webview-api';

import type { AppState, ServerSideApi, ServerSideApiDef, UpdateResult } from './apiTypes';
import { log } from './logger';
import { sampleList, store } from './store';

export function createApi(connection: MessageConnection): ServerSideApi {
    const disposables: DisposableLike[] = [];
    const dispose = createDisposeMethodFromList(disposables);

    const api: ServerSideApiDef = {
        serverRequests: {
            whatTimeIsIt,
            updateAppState,
            getAppState,
            resetTodos,
        },
        serverNotifications: {
            async showInformationMessage(message) {
                await window.showInformationMessage('Show Message: ' + message);
            },
        },
        clientRequests: {},
        clientNotifications: { onChangeAppState: true },
    };

    const serverSideApi = createServerSideHelloWorldApi(connection, api);
    disposables.push(serverSideApi);
    disposables.push(
        store.state.subscribe((v) => {
            setLogLevel(v.logLevel);
            serverSideApi.clientNotification.onChangeAppState(v);
        }),
    );

    return injectDisposable({ ...serverSideApi }, dispose);

    /**
     * Get the time
     */
    async function whatTimeIsIt() {
        return new Date().toString();
    }

    /**
     * Update the todo list
     */
    async function updateAppState(todos: AppState): Promise<UpdateResult<AppState>> {
        let success = false;

        function update(current: AppState): AppState {
            if (current.seq !== todos.seq) return current;
            const next = { ...todos };
            ++next.seq;
            success = true;
            return next;
        }

        log('Update Todos: %o', todos);
        store.state.update(update);
        return { success, value: store.state.value };
    }

    /**
     * Fetch the todo list
     */
    async function getAppState() {
        const v = store.state.value;
        log('getTodos, found: %o', v);
        return v;
    }

    /**
     * Reset the Todo list
     */
    async function resetTodos() {
        const current = store.state.value;
        updateAppState({ ...current, todos: sampleList.map((todo) => ({ ...todo })) });
    }
}
