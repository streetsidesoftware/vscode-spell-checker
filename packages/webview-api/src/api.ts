import type {
    ApplyNotificationAPI,
    ApplyRequestAPI,
    ClientAPIDef,
    ClientSideMethods,
    Logger,
    MessageConnection,
    RpcAPI,
    ServerAPIDef,
    ServerSideMethods,
} from 'json-rpc-api';
import { createClientApi, createServerApi } from 'json-rpc-api';

import type { RequestResult, Settings, SetValueRequest, SetValueResult, TextDocumentRef, TodoList, WatchFieldList } from './apiModels';

/** Requests that can be made to the extension */
export interface ServerRequestsAPI {
    whatTimeIsIt(): string;
    setLogDebug(enable: boolean): boolean;
    getLogDebug(): boolean;
    getTodos(): RequestResult<TodoList>;
    getCurrentDocument(): RequestResult<TextDocumentRef | null>;
    getDocSettings(docUrl?: string): Settings | null;
    resetTodos(): SetValueResult<TodoList>;
    setTodos(req: SetValueRequest<TodoList>): SetValueResult<TodoList>;
    watchFields(req: WatchFieldList): void;
}

/** Notifications that can be sent to the extension */
export interface ServerNotificationsAPI {
    showInformationMessage(message: string): void;
    openTextDocument(url: string): void;
}

/**
 * Requests that can be made from the extension to the webview or webviews
 * Note: RPC requests to the client/webview is rare.
 */
export interface ClientRequestsAPI {}

/** Notifications from the extension to the webview. */
export interface ClientNotificationsAPI {
    onStateChange(change: WatchFieldList): void;
}

export interface SpellInfoWebviewAPI extends RpcAPI {
    serverRequests: ApplyRequestAPI<ServerRequestsAPI>;
    serverNotifications: ApplyNotificationAPI<ServerNotificationsAPI>;
    clientRequests: ApplyRequestAPI<ClientRequestsAPI>;
    clientNotifications: ApplyNotificationAPI<ClientNotificationsAPI>;
}

/**
 * Used on the server side (in the extension) to communicate with the webviews.
 */
export interface ServerSideApi extends ServerSideMethods<SpellInfoWebviewAPI> {}
/**
 * Used in the webviews to communicate with the extension.
 */
export interface ClientSideApi extends ClientSideMethods<SpellInfoWebviewAPI> {}

export type ServerSideApiDef = ServerAPIDef<SpellInfoWebviewAPI>;
export type ClientSideApiDef = ClientAPIDef<SpellInfoWebviewAPI>;

export function createServerSideSpellInfoWebviewApi(
    connection: MessageConnection,
    api: ServerAPIDef<SpellInfoWebviewAPI>,
    logger: Logger | undefined,
): ServerSideApi {
    return createServerApi(connection, api, logger);
}

export function createClientSideSpellInfoWebviewApi(
    connection: MessageConnection,
    api: ClientAPIDef<SpellInfoWebviewAPI>,
    logger: Logger | undefined,
): ClientSideApi {
    return createClientApi(connection, api, logger);
}
