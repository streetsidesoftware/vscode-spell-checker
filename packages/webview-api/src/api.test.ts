import type { MessageConnection } from 'json-rpc-api';
import { createDisposable } from 'utils-disposables';
import { describe, expect, test, vi } from 'vitest';

import type { ClientSideApiDef } from './api';
import * as api from './api';

describe('api', () => {
    test('api', () => {
        expect(Object.keys(api).sort()).toMatchSnapshot();
    });

    test('Creating a Server API', () => {
        const connection = getConnection();
        const clientInterface: ClientSideApiDef = {
            clientNotifications: {
                onStateChange: true,
            },
            clientRequests: {},
            serverNotifications: {
                showInformationMessage: true,
                openTextDocument: true,
            },
            serverRequests: {
                getCurrentDocument: true,
                getDocSettings: true,
                getLogDebug: true,
                getTodos: true,
                resetTodos: true,
                setLogDebug: true,
                setTodos: true,
                watchFields: true,
                whatTimeIsIt: true,
            },
        };
        const client = api.createClientSideSpellInfoWebviewApi(connection, clientInterface, undefined);
        expect(client).toBeDefined();
        expect(client.serverRequest.getTodos).toBeTypeOf('function');
        const fn: (...p: any) => any = client.serverRequest.setTodos;
        expect(fn).toBeTypeOf('function');
        const methods = Object.entries(client).flatMap(([key, value]) =>
            Object.entries(value).map(([k, v]) => [key + '.' + k, v] as const),
        );
        const methodMap = Object.fromEntries(methods);
        expect(Object.keys(methodMap).sort()).toMatchSnapshot();
        expect(Object.values(methodMap).map((fn) => typeof fn)).toMatchSnapshot();
    });
});

function getConnection(): MessageConnection {
    const c: MessageConnection = {
        onNotification: vi.fn(() => createDisposable(vi.fn())),
        onRequest: vi.fn(() => createDisposable(vi.fn())),
        sendNotification: vi.fn(() => Promise.resolve()),
        sendRequest: vi.fn((() => Promise.resolve()) as () => Promise<any>),
    };
    return c;
}
