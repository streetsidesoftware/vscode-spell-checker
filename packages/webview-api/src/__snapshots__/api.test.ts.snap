// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api > Creating a Server API 1`] = `
[
  "clientNotification.onStateChange",
  "serverNotification.openTextDocument",
  "serverNotification.showInformationMessage",
  "serverRequest.getCurrentDocument",
  "serverRequest.getDocSettings",
  "serverRequest.getLogDebug",
  "serverRequest.getTodos",
  "serverRequest.resetTodos",
  "serverRequest.setLogDebug",
  "serverRequest.setTodos",
  "serverRequest.watchFields",
  "serverRequest.whatTimeIsIt",
]
`;

exports[`api > Creating a Server API 2`] = `
[
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "function",
  "object",
]
`;

exports[`api > api 1`] = `
[
  "createClientSideSpellInfoWebviewApi",
  "createServerSideSpellInfoWebviewApi",
]
`;
